pkg/provisioner/ironic/ironic.go

```diff
@@ -321,6 +321,52 @@ func (p *ironicProvisioner) getNode() (*nodes.Node, error) {
 	}
 }
 
+// Verifies that node has port assigned by Ironic.
+func (p *ironicProvisioner) nodeHasAssignedPort(ironicNode *nodes.Node) bool {
+	opts := ports.ListOpts{
+		Fields: []string{"node_uuid"},
+	}
+
+	opts.NodeUUID = ironicNode.UUID
+
+	pager := ports.List(p.client, opts)
+	if pager.Err != nil {
+		p.log.Info("node has no assigned port", "error", pager.Err)
+		return false
+	}
+
+	allPages, err := pager.AllPages()
+	if err != nil {
+		p.log.Info("node has no assigned port", "error", err)
+		return false
+	}
+
+	ok, err := allPages.IsEmpty()
+	if err != nil {
+		p.log.Info("node has no assigned port", "error", err)
+		return false
+	}
+
+	if ok {
+		p.log.Info("node has no assigned port")
+		return false
+	}
+
+	p.log.Info("node has assigned port")
+	return true
+}
+
+// Verify that MAC is already allocated to some node port.
+func (p *ironicProvisioner) isPortAllocatedToAddress(address string) bool {
+	if allPorts, err := p.listAllPorts(address); err == nil && len(allPorts) > 0 {
+		p.log.Info("address is allocated to port", "address", address)
+		return true
```

Comment:
Why not check which node the port is assigned to here, instead of listing the ports again in nodeHasAssignedPort?

Reply:
From `ports.*` functions signatures one can assume that Ironic can assign multiple ports per single node.

So to avoid hard codding logic that in our case only sigle port can be allocated to node, I've split this check into two methods.

 - `isPortAllocatedToAddress` to see if MAC is used in any of available ports
 - `nodeHasAssignedPort` to see if node has some port assigned to it, in theory it can be some other none PXE port
 
 If we are sure that one-on-one mapping of MAC to Node would not potentially break something in that future, I could refactor this PR, but this refactor is basically, `redo PR from scratch`.

Reply:
> in theory it can be some other none PXE port

This is the part I disagree with. The only port we add manually on creation of the Node is the PXE port. Therefore if that iniitial port creation fails we always need to try it again to get consistent results, regardless of the existence of any other ports.

The only scenario I can think of in which other ports exist but the PXE port does not is if the MAC address provided is wrong and inspection consequently [removes](https://github.com/metal3-io/ironic-inspector-image/blob/master/ironic-inspector.conf.j2#L36) it while adding the other ports. This doesn't seem to be related to the failure you described. With drivers where NeedsMAC() is true (i.e. the ones where the missing port is a problem) this scenario can't happen, because inspector can't find the Node to update without the PXE port existing.